pandas/tslib.pyx

```diff
@@ -1154,8 +1154,10 @@ cdef convert_to_tsobject(object ts, object tz, object unit):
         # Keep the converter same as PyDateTime's
         ts = datetime.combine(ts, datetime_time())
         return convert_to_tsobject(ts, tz, None)
-    else:
+    elif getattr(ts, '_typ', None) == 'period':
         raise ValueError("Cannot convert Period to Timestamp unambiguously. Use to_timestamp")
+    else:
```

Comment:
It's a long story. Ultimately the issue is that `SparseSeries.__getitem__` relies on the questionable assumption that non-scalar input like `slice(2)` will raise a `TypeError` when used with `Index.get_loc`. My refactor to `DatetimeIndex.get_loc` meant that these started raising `KeyError`, because I needed to swallow earlier `TypeError` that arise in `get_loc` with `method='nearest'` (e.g., when attempting to use the original label when it's a string, not a Timestamp).

I'll add a test to `DatetimeIndex.get_loc` to verify that it raises TypeError for invalid input. `SparseSeries.__getitem__` could also use a refactor to rely less on exceptions for data flow but I'd rather not touch it if possible.